linfa-bayes/src/error.rs

```diff
@@ -0,0 +1,46 @@
+use std::fmt;
+
+use ndarray::ShapeError;
+use ndarray_stats::errors::MinMaxError;
+
+pub type Result<T> = std::result::Result<T, BayesError>;
+
+#[derive(Debug)]
+pub enum BayesError {
+    /// Error when accessing an untrained model
+    UntrainedModel(String),
```

Comment:
it should not be possible to run in a situation where an error is thrown because a model is not initialized. Let me example why I think so:
 * each model should have two structs, one where the training hyperparameters are stored, and one which contains a fitted model
 * the function `Parameters::fit` should construct a fitted model from a set of hyperparameters
 * the function `Parameters::fit_with` should take a `Option<Model>` and return a new model with an updated state

A training loop would look then like this:
```
let hparams = GNBayes::params();
create_data().iter()
    .fold(None, |model, data| Some(hparams.fit_with(model, data)?))
```
or
```
let hparams = GNBayes::params();
let mut model = None;
for data in create_data() {
     model = Some(hparams.fit_with(model, data)?);
}
```
what do you think?

Reply:
I like the functional style, might require involved refactoring, will start working on it tomorrow.

Reply:
I took a stab at it today, please look at [example](https://github.com/VasanthakumarV/linfa/blob/feature/bayes/linfa-bayes/examples/winequality.rs#L57) to see how the incremental learning api changed.

I had to duplicate one parameter ([`priors`](https://github.com/VasanthakumarV/linfa/blob/feature/bayes/linfa-bayes/src/gaussian_nb.rs#L91)) between the two structs, which I don't feel good about.

Also do let me know what naming convention I should follow, for the structs and methods.

I think having one struct encompassing all methods (fit, fit_with and predict) will give the users the best experience.
Are there benefits to separating the predict function from training?

Reply:
> I took a stab at it today, please look at [example](https://github.com/VasanthakumarV/linfa/blob/feature/bayes/linfa-bayes/examples/winequality.rs#L57) to see how the incremental learning api changed.
> 
 * new should be named `default_params` and `unfitted` to give an indication what they are doing
 * the `partial_fit` could consume a model and produce one (for me its simpler to reason about the behaviour)

Another point are labels. At the moment you're reserving an additional field to `partial_fit` for them, but I think they should be implicitly stored in the targets. One way to solve this, is to have a trait for targets and a corresponding function for labels. For a Vec/Array1 this performs the `unique` function. A special struct `TargetsWithLabels` would wrap a Vec/Array1 and its corresponding labels. In this case the function would return the provided labels (and perhaps performs a check operation to see that they are complete) I have experimented with this [here](https://github.com/rust-ml/linfa/pull/45/files#diff-1b4495edaff386d67486c019daec29fdR17). After we merged the PR and figured out how the API should look like, I would like to update the syntax.

> I had to duplicate one parameter ([`priors`](https://github.com/VasanthakumarV/linfa/blob/feature/bayes/linfa-bayes/src/gaussian_nb.rs#L91)) between the two structs, which I don't feel good about.
> 
well they are different, the first one is a prior prior, and the second one an estimated prior :D 

> Also do let me know what naming convention I should follow, for the structs and methods.
> 
mh don't know, I think `partial_fit` should be called `update` or `fit_with` when the prior model is optional

> I think having one struct encompassing all methods (fit, fit_with and predict) will give the users the best experience.
> Are there benefits to separating the predict function from training?

yes that you can't produce untrained models. If you have `Model::params -> Params`, `Params::fit -> Model`, `Params::fit_with -> Model` only one struct is visible to the user, the separation is only an implementation detail
```rust
let model = GaussianNB::params()
   .with_priors(..)
   .var_smoothing(0.5)
   .fit(&x, &y, &labels);
```

Reply:
Thank you for the detailed comments,
- `new` methods in the [example](https://github.com/VasanthakumarV/linfa/blob/feature/bayes/linfa-bayes/examples/winequality.rs#L57) is now `default` and `unfitted`
- `partial_fit` is now `fit_with`
- Do we come back to fix the labels/classes argument in the `fit_with` method after merging the traits PR?
- I still had to use the `UntrainedModel` error because `unfitted` method returns a model object, using Option to avoid this isn't very elegant, because we have to return a model object wrapped in Option and eventually the user has to unwrap it

Reply:
> Thank you for the detailed comments,
> 
>     * `new` methods in the [example](https://github.com/VasanthakumarV/linfa/blob/feature/bayes/linfa-bayes/examples/winequality.rs#L57) is now `default` and `unfitted`
> 
>     * `partial_fit` is now `fit_with`
> 
>     * `fit_with` consumes a model and produces one
> 
okay :+1: 

>     * Do we come back to fix the labels/classes argument in the `fit_with` method after merging the traits PR?
> 
yes definitely, I'm playing around atm how the new traits and SVM classifier are interacting, if you want to speed the process up, do a review of the PR. 

>     * I still had to use the `UntrainedModel` error because `unfitted` method returns a model object, using Option to avoid this isn't very elegant, because we have to return a model object wrapped in Option and eventually the user has to unwrap it
yeah that is true, either its an external error or an internal one .. once we have proper error handling in the top level, a `UntrainedModel` error is due

I will look into the remaining parts tomorrow again, and when everything is fine, merge the PR